### ISSUE TITLE
Asynchronous querying

### ISSUE THREAD

**Issue by nl095059** [olifolkerd/tabulator#3136]
*Is your feature request related to a problem? Please describe.**
Some web services can take a while to collect queried data and as such have an asynchronous mechanism to allow the user to initiate the query, obtain the query status and fetch results.
Tabulator is brilliant for displaying an existing "full" set of data and paging over it but doesn't provide the ability to view "partial" datasets. It would be nice to have the ability to view an incomplete set of data, while waiting for the query to complete. 

**Describe the solution you'd like**
It would be nice to add a callback to Tabulator so that it knows a query has been (re-)initiated. It would also need to be able to construct a status query and parse the results (Similar to that of the ajaxURLGenerator/ajaxResponse for the results) which would update the pagination/counts and status. A configurable status bar would also be needed to update the user with the progress of the query.
Serious consideration would need to be given around filtering/sorting and grouping of data and whether these functions could be supported if this mode was enabled.

**Describe alternatives you've considered**
It is possible to partially implement this functionality outside of tabulator by calling into tabulator to update page counts, attaching status bars, etc. but it does feel like it would be nicer to implement this as an optional core feature.

**Comment by olifolkerd** [comment]
Tabulator already has Progressive Ajax Loading functionality built in.

It comes with two modes:
- **Load** - Loads the data in subsequent query's until all data is loaded
- **scroll** - Loads the data as the user scrolls down the table

Full details can be found in the [Progressive Ajax Loading Documentation](http://tabulator.info/docs/4.9/data#ajax-progressive)

Cheers

Oli :)

**Comment by nl095059** [comment]
Hi Oli,

  I did look into the progressiveLoad modes, but the documentation states that this doesn't work with the pagination controls. Looking back at my original request, I think I made it a little too generic and didn't focus enough on the fact that we have the requirement to use a pagination control for this functionality. The other thing I I didn't make clear is that due to the backend technologies we are using the number of results aren't known at the point of requesting the first page of results! We therefore need to poll for the query status seperately from the result page fetching.

This would make this requirement less of a "progressive" load and more of an ad-hoc (Load the specified page when you need it mode). There would then be the need to check the status of the query and update the pagination control based on the "current" status of the query as the number of results might not be known when the initial results come back.

Please let me know if this is something you want to see if tabulator, or if individual elements of it would be useful?

Cheers

**Comment by olifolkerd** [comment]
Hey @nl095059 

Im afraid i still dont understand what you are looking for, the pagination system in remote mode will request data from the server one page at a time as the user wants to view that page, there is no need to progressively load anything in that scenario, your page size should be set to be small enough that it can be quickly returned from the server.

As for not knowing the page count on the initial load, simply return a page count of 1 on the first request, then you can call the `setData` function at any point later to rerun the request and return the correct page count.

Cheers

Oli

**Comment by nl095059** [comment]
Hi Oli,

  It sounds like you are describing what we want, but looking at the documentation [](http://tabulator.info/docs/4.9/data#ajax-progressive) it states that _'This mode can be enable using the ajaxProgressiveLoad option. No pagination controls will be visible on screen, it just reusues the functionality of the pagination module to sequentially load the data.'_.

If I try and enable both progressiveMode: 'load' and pagination: 'remote' then the following message gets thrown from ajax.js:66 'Progressive Load Error - Pagination and progressive load cannot be used at the same time'

I really hope I am missing something and that adding a simple flag will fix it :-)

Cheers

**Comment by olifolkerd** [comment]
I think some wires are getting crossed here

My point is that you don't need to enable progressive loading when using remote pagination, it automatically makes a request for each page as the user navigates to it. the two modes are mutually exclusive, they are both methods of requesting data from the server once at a time, the pagination mode paginates the data, the progressive loading renders it as a scrolling list. They are two totally different ways of loading/displaying data in the table.

Cheers

Oli

**Comment by nl095059** [comment]
Hi Oli,

  Sorry, I've been off for a few weeks over Christmas and struggling to articulate this problem!  The progressiveLoad documentation has thrown me off a bit, as we definately want to use remote pagination!

The way I understand it at present:

1. We can use remote pagination to fetch each page as the user requests it. 
2. The first page can set an artificial page count of 1 (Just to load the first page and render the controls).
3. Currently an external process will need to poll for the query status (To update the pagination controls page count with more pages as it is "processing")


The main extensions we would need are largely as articulated in the original ask:

> It would be nice to add a callback to Tabulator so that it knows a query has been (re-)initiated. It would also need to be able to construct a status query and parse the results (Similar to that of the ajaxURLGenerator/ajaxResponse for the results) which would update the pagination/counts and status. A configurable status bar would also be needed to update the user with the progress of the query.
> Serious consideration would need to be given around filtering/sorting and grouping of data and whether these functions could be supported if this mode was enabled.
> 

Does this make more sense? I can knock together a sequence diagram together if that would help clean things up?

**Comment by olifolkerd** [comment]
This sounds like an unusual edge case, and is not functionality i would be looking to add into core Tabulator.

If you need this level of bespoke functionality then it is correct that this sites outside of Tabulator.

As it stands all you need to do is make an initial request to pull down the first page (this can happen through tabulator using the normal remote pagination setup), then using your external monitoring process *(which should totally be outside of Tabulator, this is so bespoke to the usage case that to integrate into Tabulator would be an anit-pattern)* check when the system is ready, at that point if you call the `setPage` or `setData` function on the table it will rerun the request to the server which should return the correct page_count variable this time, this will result in the correct number of page controls being created in the footer. if you want to change the url for this second request then just pass the new url into the `setData` function.

Any form of graphical overlay you would like to include during this will need to be handled outside of Tabulator.

If you want to override the existing ajax request system to handle things in a more complex fashion then you can use the `ajaxRequestFunc` property to replace the ajax promise it will be called when ever Tabulator needs data and should return a promise that resolves when the data is available. This should also show a "loading" popup when loading data. More info on this can be found in the [Ajax Documentation](http://tabulator.info/docs/4.9/data#ajax-promise)

Cheers

Oli :)

**Comment by nl095059** [comment]
Hi Oli,

  Thanks for taking the time to understand our use case. I completely agree that this asynchronous query setup/pagination is an unusual use case for the majority of tabulators users. Sadly it's one of many patterns we have been working with for many years in our domain which sometimes makes using open source a little tricky without clunky wrappers/mods.

I will implement something outside of our codebase, but also off to the side of tabulator so other teams here might be able to utilise it too.

Cheers

**Comment by nl095059** [comment]
Hi Oli,

  I realise this is not the best place for this, but struggling to find a
more acceptable place.

I am doing a little tidying of the Ajax and Page modules to allow me write
some tests and to understand the actual functionality clearer. I have
realised that a quick win for clarity is to move the HTML Components from
these modules into seperate classes to split the business logic from the
presentation (While still residing in the same js file). Are you willing to
entertain a couple of PR's for this work? And if so, would you like
seperate issue(s) raised for traceability/external visibility?

Cheers

**Comment by olifolkerd** [comment]
Hey @nl095059 

The entire library is currently being completely refactored into classes as part of the 5.0 rebuild, to allow the introduction of unit tests for every part of the system. It is also breaking down each of the modules into a series of smaller files as each of the module files has now grown to unmanageable sizes.

For this reason i am not currently accepting any PRs of the 4.9 code base as they will not be compatible with the new code structure.

I would recommend holding off any changes for a couple of weeks until i have finished the refactor, i would then be open to any feedback that you may have. If it involves refactoring modules then please create an issue before starting a PR so we can ensure we are on the same page. 

Also please note that all modules will need to follow the same format, so i would not accept a PR that changes the internal structure of only one or two modules if this causes them to deviate from the others as it will make the codebase harder to maintain and difficult for other developers.

If there are any particular areas you are struggling with, could you create a new issue to discuss them and I will see what I can include as part of my refactor. Though I must be honest there isn't much HTML generation in the pagination and AJAX and pagination modules *(only button and placeholder generation)* and i feel that trying to pull these out of the module would make things overly complicated to maintain with little benefit actually arising from the change, so i would need to see a strong case for why you think this should occur.

Cheers

Oli
